image-rs/src/registry/mod.rs

```diff
@@ -0,0 +1,494 @@
+                })
+                .max_by_key(|&(matched_len, _)| matched_len);
+
+            // if no rules are found. Just pull the original task.
+            let Some((matched_len, registry_rule)) = matched_rule else {
+                let original_task = ImagePullTask {
+                    image_reference: task.image_reference,
+                    use_http: false,
+                };
+                tasks.push(original_task);
+                continue;
+            };
+
+            // check if the pulling is blocked
+            if registry_rule.blocked {
+                debug!("Image {reference_string} is blocked by registry rule.");
+                continue;
+            }
+
+            // Firstly, Mirror tasks
+            for mirror in &registry_rule.mirror {
+                let mirrored_reference =
+                    format!("{}{}", mirror.location, &reference_string[matched_len..]);
+                let task = ImagePullTask {
+                    image_reference: Reference::from_str(&mirrored_reference)?,
+                    use_http: mirror.insecure,
+                };
+
+                tasks.push(task);
+            }
+
+            // Then, Remapping task
+            if !registry_rule.location.is_empty() {
+                let remapping_reference = format!(
+                    "{}{}",
+                    registry_rule.location,
+                    &reference_string[matched_len..]
+                );
+                let remapping_task = ImagePullTask {
+                    image_reference: Reference::from_str(&remapping_reference)?,
+                    use_http: registry_rule.insecure,
+                };
+                tasks.push(remapping_task);
+            }
+        }
+
+        Ok(tasks)
+    }
+
+    /// Due to the [Remapping and mirroring registries](https://github.com/containers/image/blob/main/docs/containers-registries.conf.5.md#remapping-and-mirroring-registries),
+    /// `mirror` will let the client to try many different image
+    /// sources, thus this function will return a list of image pull tasks for future
+    /// use.
+    pub fn process(&self, image_reference: Reference) -> Result<Vec<ImagePullTask>> {
+        let mut tasks = Vec::new();
+
+        // First, try all unqualified search registries
+        //
+        // The original rule is to add registries to handle an unqualified image.
+        // Crate `oci-spec`'s parser logic of [`oci_client::Reference`] will automatically
+        // add `docker.io` to unqualified images.
+        // Thus this call will try to add tasks for image pull request whose registry
+        // is `docker.io`.
+        //
+        // TODO: This is sometimes not a good idea, as the user might want to
+        // pull images like `docker.io/foo/bar` directly but current implementation will add
+        // tries from registries in `unqualified_search_registries` We can handle this later.
+        if image_reference.registry() != "docker.io" {
```

Comment:
Would it help the above to check if the registry starts with "docker.io" rather than exactly equaling it?

Reply:
The exactly equaling will help to avoid cases where registry is like `docker.io.com` rather than `docker.io`. The source code to add `docker.io` is [here](https://github.com/youki-dev/oci-spec-rs/blob/main/src/distribution/reference.rs#L310)